src/ucp/core/ucp_request.c

```diff
@@ -273,6 +274,101 @@ int ucp_request_pending_add(ucp_request_t *req, unsigned pending_flags)
               ucs_status_string(status));
 }
 
+static unsigned ucp_request_dt_invalidate_progress(void *arg)
+{
+    ucp_request_t *req = arg;
+
+    ucp_request_complete_send(req, req->send.state.uct_comp.status);
+    return 1;
+}
+
+static void ucp_request_dt_invalidate_cb(void *arg)
+{
+    ucp_request_t *req = arg;
+
+    ucp_invoke_uct_completion(&req->send.state.uct_comp,
+                              req->send.state.uct_comp.status);
+}
+
+static void ucp_request_err_uct_comp_cb(uct_completion_t *comp)
+{
+    ucp_request_t *req         = ucs_container_of(comp, ucp_request_t,
+                                                  send.state.uct_comp);
+    uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    uct_worker_progress_register_safe(req->send.ep->worker->uct,
+                                      ucp_request_dt_invalidate_progress,
+                                      req, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                      &prog_id);
+}
+
+static void ucp_request_dt_invalidate(ucp_context_t *context,
+                                      uct_mem_h *uct_memh,
+                                      ucp_md_map_t *md_map_p,
+                                      ucp_request_t *req)
+{
+    uct_md_mem_dereg_params_t params = {
+        .field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH |
+                      UCT_MD_MEM_DEREG_FIELD_FLAGS |
+                      UCT_MD_MEM_DEREG_FIELD_CALLBACK |
+                      UCT_MD_MEM_DEREG_FIELD_ARG,
+        .flags      = UCT_MD_MEM_DEREG_FLAG_INVALIDATE,
+        .cb         = ucp_request_dt_invalidate_cb,
```

Comment:
we missed it but UCT API should accept uct_completion_t* as parameter for uct_md_mem_dereg_v2 and not callback+arg

Reply:
we discussed this API on meeting: using UCT completion here will require including UCT headers into MD (or use forward declaration). is this ok from design perspective?

Reply:
i'd expect UCT headers will already be included from UCT MD code (note: this is not same as rcache)